### DISCUSSION TITLE
Fixed Baseline Distance Paragraph Height

### DISCUSSION THREAD

**Discussion by SanjayTml** [sile-typesetter/sile#1458] [Q&A] [answered]
I have been working on creating posters for different fonts using SILE and Lua. When I try to calculate the paragraph height after changing the linespacing method to fixed and providing a fixed baselinedistance value, I can not get the correct value.

I use the following lines of code to calculate the paragraph height:

`

    SILE.call("font", {size = fontOptions.size * relratio}, function()

                SILE.process({name})                                                 -- name is the paragraph text
                SILE.typesetter:leaveHmode(1)
                subsidiary = SILE.pagebuilder:collateVboxes(SILE.typesetter.state.outputQueue)
                local queue = SILE.typesetter.state.outputQueue
                for i = 1, #queue do
                    if queue[i].is_vbox then
                        linecounter = linecounter + 1
                        calcparheight = calcparheight + queue[i].height + queue[i].depth
                    end
                end
                SILE.typesetter:pushState()
            end)

    print("Paragraph height = ", subsidiary.height)
    print("Calculated Paragraph height = ", calcparheight)

`

Here 'subsidiary.height' gives me the standard paragraph height that SILE calculates using vboxes, and 'calcparheight' gives me paragraph height with depth(ascenders & descenders) included after calculations(sum of each line height). Both of them work fine when the linespacing is off but gives incorrect paragraph height when linespacing is on.

I change the linespacing method inside Lua as below:

`

     SILE.call("linespacing-on")
     SILE.settings.set("linespacing.method",'fixed')
     SILE.settings.set("linespacing.fixed.baselinedistance", SILE.length('4.5em'))

`

I need help on fixing the above code or solving this using any other method or function in Lua or SILE to calculate the paragraph height.

**Comment by alerque** [comment]
I think your calculations using the default spacing might be wrong (but by happy accident coming in close). You do realize you cannot know the final height of a paragraph until during the page break routine when all aspects of the height are finalized, right? Anything you try to do in the flow of things like you are is going to be either a guess or you are going to fudge and force some things in the page builder to happen early and hence fix parts of the page before other parts, which can have weird side effects.

Do you have a more complete working example I can look at for where you are calculating this and what gets done with it afterwards?

**Reply by SanjayTml** [reply]
Thank you for the response. I have two files, one poster.lua and one text.sil

Poster.lua looks like:
```
local plain = SILE.require("classes/plain")
local poster = plain { id = "poster"}

function poster:init()
    SILE.settings.set("document.parindent", SILE.nodefactory.glue(0))
    return plain.init(self)
end

SILE.registerCommand("posterOutput", function(options, content)
    mystring = "This is Lorem Ipsum random paragraph. This paragraph should occupy the largest possible height of a single page without overflowing to another page. Thank you!"

    SILE.call("linespacing-on")
    SILE.settings.set("linespacing.method",'fixed')
    setmaxheight(mystring)
end)

-- Outputs the string in maximum height available for single page
function setmaxheight(str, setHeight)
    local fontOptions = SILE.font.loadDefaults({})
    local name = str
    if setHeight == nil then
        setHeight = SILE.measurement("95%fh"):absolute()
    end
    if name then
        local tokens = SILE.shaper:shapeToken(name,fontOptions)
        local w = 0
        local h = 0
        for j= 1,#tokens do
            w = w + tokens[j].width
            if h < tokens[j].height then
                h = tokens[j].height
            end
        end
        local fratio = 70/100 * SILE.scratch.pagewidth / 60/100 * SILE.scratch.pageheight
        local cratio = tokens[1].height / tokens[1].width
        local relratio = fratio/cratio
        local subsidiary = nil
        local parheight = nil
        local parwidth = nil
        local linecounter = 0
        local calcparheight = 0
        local largestlineheight = 0
        local smallestlineheight = 1000
        local currentlineheight = 0
        local l = SILE.length({ length = 20, stretch = 10, shrink = 10})
        table.remove(SILE.typesetter.state.nodes)
        local i = 0
        
        -- Initial parheight calculation with default values
        SILE.call("font", {size = fontOptions.size * relratio}, function()
                SILE.process({name})
                SILE.typesetter:leaveHmode(1)
                subsidiary = SILE.pagebuilder:collateVboxes(SILE.typesetter.state.outputQueue)
                local queue = SILE.typesetter.state.outputQueue
                for i = 1, #queue do
                    if queue[i].is_vbox then
                        linecounter = linecounter + 1
                        calcparheight = calcparheight + queue[i].height + queue[i].depth
                        currentlineheight = queue[i].height + queue[i].depth
                        if(currentlineheight > largestlineheight) then
                          largestlineheight = currentlineheight
                        end
                        if(currentlineheight ~= SILE.length('0pt'):absolute() and currentlineheight < smallestlineheight) then
                          smallestlineheight = currentlineheight
                        end
                    end
                end
                SILE.typesetter:pushState()
            end)
        local defaultlinespacing = SILE.settings.get("linespacing.fixed.baselinedistance"):absolute()
        local baselinedistancevalue = defaultlinespacing
        repeat
            local totalHeight = 0
            table.remove(SILE.typesetter.state.nodes)
            local newfontOptions = SILE.font.loadDefaults({size = fontOptions.size*relratio})
            local newtokens = SILE.shaper:shapeToken(str,newfontOptions)
            local wprime = 0
            local hprime = 0
            for k= 1, #newtokens do
                wprime = wprime + newtokens[k].width
                if newtokens[k].height > hprime then
                    hprime = newtokens[k].height
                end
            end
            baselinedistancevalue = SILE.length('4.5em'):absolute() + largestlineheight - hprime
            SILE.settings.set("linespacing.fixed.baselinedistance", baselinedistancevalue)
            parheight = calcparheight
            
            -- Adjust the font relratio(size)
                if(parheight < 50/100*setHeight) then
                    relratio = relratio * 1.75
                elseif(parheight < 80/100*setHeight) then
                    relratio = relratio * 1.25
                elseif(parheight < 90/100*setHeight) then
                    relratio = relratio * 1.05
                elseif(parheight > 1.25*setHeight) then
                    relratio = relratio * 0.7
                elseif(parheight > 95/100*setHeight) then
                    relratio = relratio * 0.9
                else
                    break
                end
                
            calcparheight = 0
            linecounter = 0
            smallestlineheight = largestlineheight + 1000
            currentlineheight = 0
            largestlineheight = 0
            subsidiary = nil
            table.remove(SILE.typesetter.state.nodes)

            -- Parheight Calculation with adjusted relratio(font size)
            SILE.call("font", {size = fontOptions.size * relratio}, function()
                    SILE.process({name})
                    SILE.typesetter:leaveHmode(1)
                    subsidiary = SILE.pagebuilder:collateVboxes(SILE.typesetter.state.outputQueue)
                    local queue = SILE.typesetter.state.outputQueue
                    for i = 1, #queue do
                        totalHeight = totalHeight + queue[i].height + queue[i].depth
                        if queue[i].is_vbox then
                            linecounter = linecounter + 1
                            calcparheight = calcparheight + queue[i].height + queue[i].depth
                            currentlineheight = queue[i].height + queue[i].depth
                            if(currentlineheight > largestlineheight) then
                              largestlineheight = currentlineheight
                            end
                            if(currentlineheight ~= SILE.length('0pt'):absolute() and currentlineheight < smallestlineheight) then
                              smallestlineheight = currentlineheight
                            end
                        end
                    end
                    SILE.typesetter:pushState()
                end)
            i = i+1
            if(i > 12 and parheight < 700) then
                    baselinedistancevalue = SILE.length('4.5em'):absolute() + largestlineheight - hprime
                    SILE.settings.set("linespacing.fixed.baselinedistance", baselinedistancevalue)
                    parheight = calcparheight
                    break
            end
        until(false)
        print(" Final parheight = ", parheight)
        SILE.call("hbox", {height = 0})

        -- Print on the pdf page
        SILE.call('font', {size = fontOptions.size * relratio}, function()
             SILE.typesetter:typeset(name)
        end)
end
```

text.sil looks like:
```
\begin[class=poster,papersize=1080pt x 720pt]{document}

% Loading Packages
\script[src=packages/linespacing]
\script[src=packages/background]
\script[src=packages/color]
\script[src=packages/frametricks]
\script[src=packages/specimen]
\script[src=packages/grid]
\script[src=packages/rebox]

% Settings
\nofolios   % Turn off page numbering
\background[color=#00ff00,allpages=true]    % change background color

% Body
\posterOutput[]

\end{document}
```

**Comment by Omikhleia** [comment, marked as answer]
Your code is hard to guess with uncomplete script, undeclared variables, etc.

Here is a working version:
```
\begin{document}
\script[src=packages/linespacing]
\script{
SILE.settings.set("linespacing.method",'fixed')
SILE.settings.set("linespacing.fixed.baselinedistance", SILE.length('4.5em'))

local calcparheight = SILE.length()
SILE.process({ "I have been working on creating posters for different fonts using SILE and Lua. When I try to calculate the paragraph height after changing the linespacing method to fixed and providing a fixed baselinedistance value, I can not get the correct value." })
SILE.typesetter:leaveHmode(1)
local subsidiary = SILE.pagebuilder:collateVboxes(SILE.typesetter.state.outputQueue)
local queue = SILE.typesetter.state.outputQueue
for i = 1, #queue do
   calcparheight = calcparheight + queue[i].height + queue[i].depth
end

print("Paragraph height = ", subsidiary.height + subsidiary.depth)
print("Calculated Paragraph height = ", calcparheight)
}
\end{document}
```
Outputs:
```
Paragraph height = 	55.009765625pt
Calculated Paragraph height = 	55.009765625pt
```

With both giving the same results, which is logical since the loop is (more or less) what the collateVboxes does (wrapping the vboxes in a single vbox, which will evaluate its height (see the implementation of the `append` method of `nodefactory.vbox`).

Note the results are the same with the beginning of the script being:
```
%\script[src=packages/linespacing]
\set[parameter=document.baselineskip,value=4.5em]
\set[parameter=document.lineskip,value=0]
\script{
--SILE.settings.set("linespacing.method",'fixed')
--SILE.settings.set("linespacing.fixed.baselinedistance", SILE.length('4.5em'))
```
(I.e. the default TeX linespacing but tried to make "fixed"... At least for this text only input in a single size)

Your initial code was testing on `is_vbox` so skipping the (linespacing etc.) glues -- No idea whether it is what you wanted or not...

**Reply by Omikhleia** [reply]
As an additional note, the work in progress on #1344 will provide a "paragraph box" that one can build and remove from out queue, then measure (this is how the related "ptable" work after all for assembling cells) -- but cancelling the vertical stretch/shrink if any is present.

**Reply by Omikhleia** [reply]
@SanjayTml You are perhaps making things _much_ more complex than they ought.

In the above code you provided under alerque's answer, it's hard to understand what it is supposed to do, but it does lots of weird things.
- There are three occurrences of `table.remove(SILE.typesetter.state.nodes)`, without anything obvious having been pushed to the node queue just before. Toying with the node queue  is kind of an advance technique that might be changed one day; anyhow the common use case is to do it after a box has been _just_ pushed, to remove it, e.g. for measuring or hacking it. The most common pattern, that is, usually looks like:
  ```
  local hbox = SILE.call("hbox", {}, ...something...)
  table.remove(SILE.typesetter.state.nodes) -- steal it back from queue
  ```
  It's already, in this form, not so neat... -- but you code calls the second line without clear context. I would recommend trying to avoid it at all cost.
- Your code also does several `SILE.typesetter:pushState()`... This usually goes in pair with a corresponding `SILE.typesetter:popState()` (it's a stack of states), whereas I see none here.
- There are other weird things (e.g. near the end `SILE.call("hbox", {height = 0})` - why is there such an empty box?) which make it unclear which problem you were trying to solve.
- It's not a minimum script... (why create a class? why showing us a SIL file with green background, plenty of unused packages, etc.). Simplify!
- It has plenty of hard-coded unexplained number values (some `1000`, some `i > 12`, mind-boggling `relratio` logic...)

More generally, Your initial problem, as stated ("This paragraph should occupy the largest possible height of a single page without overflowing to another page") has also some issue of it own:
- Do you expect the text to be justified? Because, as you surely understand, justification relies on stretching/shrinking spaces, so the size computation is not that obvious...
- Do you expect the text to be hyphenated? Likewise, it introduces dashes...

I am afraid it might sound blunt, but the above code, IMHO, is **not** fixable. You'd need to rethink both the problem and the algorithm...

Here is my take at this _revisited_ challenge - each page is a "poster":
![image](https://user-images.githubusercontent.com/18075640/176983473-7000b9e1-7e82-4984-a557-890409630e08.png)

I did my homework. Your turn! :cat:

**Reply by SanjayTml** [reply]
Thank you for the hints and help, but my main aim for calculating the paragraph height is to center the paragraph vertically on a single page, along with the justification. This I couldn't figure out because I think the problem is with the paragraph height not being accurate. 
To vertically center the paragraph, I calculate the paragraph height after resizing it to the largest font size possible for the single page of 1080*720 px, then I subtract it from the remaining page height and provide a vglue to the top for half of the available space.

**Reply by alerque** [reply]
If your goal is to center a paragraph vertically on a page you don't need to muck around with calculating its height. Just put one `\vfill` before the paragraph and one after and you're done. Note that the default page-end command includes a `vfill`  already, so unless you manually eject the page all you need is `\vfill` + `your content` + `\pagebreak`.

The logic is that all the *vfill*'s will automatically get their fair portion of the leftover vertical space when the page layout happens and if there are 2 on the page each will get half the space.

**Reply by Omikhleia** [reply]
I don't get it and probably missed the point. If you size the paragraph to the largest font possible for it to fill a full page, there's not much left to center vertically ??

EDIT. "Along with justification"... There we go.
![image](https://user-images.githubusercontent.com/18075640/177822628-87cfeca0-16a0-4234-8131-f38402ae842a.png)

(Obviously one would need to use a different page layout to get really centered without different top and bottom margins and also without folio space. But that's a wholly different topic.)

**Reply by SanjayTml** [reply]
@Omikhleia Can you show/share with me how you did the above? 
I tried to edit my code but still no good results. Maybe I should start from the beginning again.

**Reply by Omikhleia** [reply]
@SanjayTml You are on a lucky day, I had kept it... Fully commented script `poster.sil` zipped and attached below. It was made for SILE 0.12.5, but it works too with SILE 0.13.2 -- except it shows some minor deprecation warnings.

[poster.zip](https://github.com/sile-typesetter/sile/files/9143562/poster.zip)

(License: CC0 public domain)

Please note the solution might not be what you initially asked - it wasn't clear, and I certainly did not use a fixed baseline. That's IMHO not doable.
Please also note the algorithm to find the ratio is not very efficient. A dumb dichotomy between ratio 1 and the estimated maximum ratio (that one is quite cleverly guessed, though).

I hope my script deserves the "Answered" status ;)

**Reply by SanjayTml** [reply]
Thank you for the code. It worked as anticipated for normal paragraphs and not when we used a fixed baseline. I have marked it answered.
